### PR TITLE
KEP-4569: Separate cgroup v1 and v2 manager implementations

### PR DESCRIPTION
**Pull request by harche**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it: 
One of the goals of cgroup v1 maintenance mode KEP is to have [separate implementation for cgroup v1 and v2](https://github.com/kubernetes/enhancements/blob/a4f67cd757eae8e00c95c0ff26eac1177febc964/keps/sig-node/4569-cgroup-v1-maintenance-mode/README.md#separation-of-cgroup-v1-and-cgroup-v2-code-paths) of cgroup manager interface. This will help us clearly distinguish the code that is in maintenance mode (cgroup v1) vs the the one that is in active development (cgroup v2).  

I tried to move as many methods as possible into the respective cgroup related implementation/files by keeping the code duplication minimum.

#### Which issue(s) this PR fixes:

Fixes # https://github.com/kubernetes/enhancements/issues/4569

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Review by kuzminxyz — APPROVED** [review]

**Comment by pacoxu** [comment]
Is this part of https://github.com/kubernetes/enhancements/issues/4569?

**Comment by harche** [comment]
@pacoxu that's correct. Specifically, this PR will try to address this point in the KEP design, 

https://github.com/kubernetes/enhancements/blob/a4f67cd757eae8e00c95c0ff26eac1177febc964/keps/sig-node/4569-cgroup-v1-maintenance-mode/README.md#separation-of-cgroup-v1-and-cgroup-v2-code-paths

**Comment by harche** [comment]
PTAL @kubernetes/sig-node-pr-reviews

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 1]
```diff
@@ -543,100 +462,19 @@ func (m *cgroupManagerImpl) ReduceCPULimits(cgroupName CgroupName) error {
...
-		file = "memory.usage_in_bytes"
-	}
-	val, err := fscommon.GetCgroupParamUint(path, file)
-	return int64(val), err
-}
-
 // Convert cgroup v1 cpu.shares value to cgroup v2 cpu.weight
 // https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2#phase-1-convert-from-cgroups-v1-settings-to-v2
-func CpuSharesToCpuWeight(cpuShares uint64) uint64 {
+func cpuSharesToCPUWeight(cpuShares uint64) uint64 {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/cgroup_manager_linux.go:467]: do we still need it here? Cannot find a callsite

**Review by SergeyKanzhelev — COMMENTED** [review]
/kind cleanup
/priority backlog
/triage accepted

just one comment. Otherwise looks good

**Review by saschagrunert — COMMENTED** [review]
Code LGTM beside @SergeyKanzhelev's question.

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const cgroupv1MemLimitFile string = "memory.limit_in_bytes"
+
+// cgroupV1impl implements the CgroupManager interface
+// for cgroup v1.
+// Its a stateless object which can be used to
+// update,create or delete any number of cgroups
```
Comment (toVersus) [inline pkg/kubelet/cm/cgroup_v1_manager_linux.go:39]:
```suggestion
// update, create or delete any number of cgroups
```

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	cgroupCommon
+}
+
+func NewCgroupV1Manager(cs *CgroupSubsystems, cgroupDriver string) CgroupManager {
+	return &cgroupV1impl{
+		cgroupCommon: newCgroupCommon(cs, cgroupDriver),
+	}
+}
+
+// Exists checks if all subsystem cgroups already exist
```
Comment (toVersus) [inline pkg/kubelet/cm/cgroup_v1_manager_linux.go:51]:
```suggestion
// Validate checks if all subsystem cgroups are valid
```

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	if len(missingPaths) > 0 {
+		return fmt.Errorf("cgroup %q has some missing paths: %v", name, strings.Join(missingPaths, ", "))
+	}
+
+	return nil
+}
+
+func (c *cgroupV1impl) Exists(name CgroupName) bool {
```
Comment (toVersus) [inline pkg/kubelet/cm/cgroup_v1_manager_linux.go:86]:
```suggestion
// Exists checks if all subsystem cgroups already exist
func (c *cgroupV1impl) Exists(name CgroupName) bool {
```

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+		return fmt.Errorf("could not read controllers for cgroup %q: %w", name, err)
+	}
+	difference := neededControllers.Difference(enabledControllers)
+	if difference.Len() > 0 {
+		return fmt.Errorf("cgroup %q has some missing controllers: %v", name, strings.Join(sets.List(difference), ", "))
+	}
+	return nil
+}
+
+func (c *cgroupV2impl) Exists(name CgroupName) bool {
```
Comment (toVersus) [inline pkg/kubelet/cm/cgroup_v2_manager_linux.go:65]:
```suggestion
// Exists checks if all subsystem cgroups already exist
func (c *cgroupV2impl) Exists(name CgroupName) bool {
```

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	cgroupCommon
+}
+
+func NewCgroupV2Manager(cs *CgroupSubsystems, cgroupDriver string) CgroupManager {
+	return &cgroupV2impl{
+		cgroupCommon: newCgroupCommon(cs, cgroupDriver),
+	}
+}
+
+// Exists checks if all subsystem cgroups already exist
```
Comment (toVersus) [inline pkg/kubelet/cm/cgroup_v2_manager_linux.go:50]:
```suggestion
// Validate checks if all subsystem cgroups are valid
```

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+	if err != nil {
+		return nil, err
+	}
+	controllers := strings.Fields(string(controllersFileContent))
+	return sets.New(controllers...), nil
+}
+
+// buildCgroupUnifiedPath builds a path to the specified name.
+func (m *cgroupCommon) buildCgroupUnifiedPath(name CgroupName) string {
+	cgroupFsAdaptedName := m.Name(name)
```
Comment (toVersus) [inline pkg/kubelet/cm/cgroup_v2_manager_linux.go:202]: Only cgroup v2 impl uses this method.

```suggestion
func (c *cgroupV2impl) buildCgroupUnifiedPath(name CgroupName) string {
	cgroupFsAdaptedName := c.Name(name)
```

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 1 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: Thanks for pointing that out. I moved `cpuSharesToCPUWeight()` and `cpuWeightToCPUShares()` in `pkg/kubelet/cm/cgroup_v2_manager_linux.go` . Those are called by `setCgroupCPUConfig()` and `getCgroupCPUConfig()` respectively in cgroup v2 implementation.

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go (hunk 2 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_v1_manager_linux.go]: Updated. Thanks.

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go (hunk 3 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_v1_manager_linux.go]: Updated. Thanks.

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go (hunk 4 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_v1_manager_linux.go:87]: Updated. Thanks.

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go (hunk 5 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_v2_manager_linux.go:66]: Updated. Thanks.

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go (hunk 6 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_v2_manager_linux.go]: Updated. Thanks.

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go (hunk 7 above)
Comment (harche) [reply pkg/kubelet/cm/cgroup_v2_manager_linux.go]: Moved it in `pkg/kubelet/cm/cgroup_v2_manager_linux.go`. Thanks.

**Comment by harche** [comment]
Thanks @rphillips @SergeyKanzhelev @toVersus for the review comments. I have updated the PR with those suggestions.

**Comment by harche** [comment]
> /lgtm
> 
> This looks ready to me.

Thanks @rphillips 

@SergeyKanzhelev whenever you get time can you review and approve it?

File: pkg/kubelet/cm/cgroup_v1_manager_linux.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const cgroupv1MemLimitFile string = "memory.limit_in_bytes"
+
+// cgroupV1impl implements the CgroupManager interface
+// for cgroup v1.
+// Its a stateless object which can be used to
```
Comment (mrunalp) [inline pkg/kubelet/cm/cgroup_v1_manager_linux.go:38]:
```suggestion
// It's a stateless object which can be used to
```

File: pkg/kubelet/cm/cgroup_v2_manager_linux.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	cmutil "k8s.io/kubernetes/pkg/kubelet/cm/util"
+)
+
+const cgroupv2MemLimitFile string = "memory.max"
+
+// cgroupV2impl implements the CgroupManager interface
+// for cgroup v2.
+// Its a stateless object which can be used to
```
Comment (mrunalp) [inline pkg/kubelet/cm/cgroup_v2_manager_linux.go:37]:
```suggestion
// It's a stateless object which can be used to
```

**Review by mrunalp — APPROVED** [review]
/lgtm

**Pull request merged**